scripts/traffic_generator/traffic_generator.py

```diff
@@ -136,44 +151,68 @@ def _filter_workers(self, workers: WorkersDict) -> WorkersDict:
-            if dataset.id in state
+            if dataset.id in datasets.keys()
         ]
         if not stored_datasets:
             logging.warning(f"Worker {worker_id} has no datasets to query")
             return None
 
         template, dataset_id, dataset_name = random.choice(stored_datasets)
-        query_url = f'{PORTAL_URL}/datasets/{dataset_id}/query/{worker_id}'
+        query_url = f'{PORTAL_URL}/datasets/{dataset_name}/finalized-stream'
         query = random.choice(self._query_templates[template])
-        query['fromBlock'], query['toBlock'] = random_range(state[dataset_id])
+        query_id = self._query_templates[template].index(query)
+        query['fromBlock'], query['toBlock'], block_start = random_range(datasets[dataset_id], chunks = NUM_CHUNKS_PER_QUERY)
 
         logging.debug(
-            f"Sending query worker_id={worker_id} query_url={query_url} "
+            f"Sending stream query_url={query_url}"
             f"from={query['fromBlock']} to={query['toBlock']}")
-        try:
-          response = requests.post(query_url, json=query, stream=True, timeout=QUERY_TIMEOUT_SEC)  # stream=True to discard response body
-        except:
-            logging.info(f"Query timed out. worker_id={worker_id}")
-            return 499
-
-
-        if response.status_code != 200:
-            logging.info(
-                f"Query failed. worker_id={worker_id} status={response.status_code} "
-                f"msg='{response.text}' elapsed={response.elapsed}")
-        else:
-            logging.info(f"Query succeeded. worker_id={worker_id} elapsed={response.elapsed}")
-        return response.status_code
+        for attempt in range(10):
+            code, ret_count, last_block = self._do_request(query_url, query)
+            logging.debug(f"attempt = {attempt} code = {code} last = {last_block}")
+            if code == 204:
+                break
+            if code == 499:
+                break
+            if ret_count == 0:
+                break
+            if last_block == query['toBlock']:
+                break
+            if last_block == -1:
+                break
+            query['fromBlock'] = last_block + 1
 
+        logging.info(f"from={query['fromBlock']} to={query['toBlock']} ret={ret_count} last_block={last_block}")
+        return code, dataset_id, block_start
 
-def random_range(ranges: List[BlockRange]) -> Tuple[int, int]:
-    r = random.choice(ranges)
-    from_block = random.randint(r.begin, r.end)
-    to_block = min(random.randint(from_block, r.end), from_block + MAX_BLOCK_RANGE)
-    return from_block, to_block
+def random_range(ranges: List[BlockRange], chunks: int) -> Tuple[int, int]:
+    first_chunk_idx = random.randint(0, len(ranges) - chunks - 1)
+    first_chunk = ranges[first_chunk_idx]
+    last_chunk = ranges[first_chunk_idx + chunks]
+    from_block = random.randint(first_chunk.begin, first_chunk.end)
+    to_block = random.randint(last_chunk.begin, last_chunk.end)
```

Comment:
If that's the only reason to read the `/state`, you can use `max_chunks` url param instead: https://github.com/subsquid/sqd-portal/commit/f2e9d758e9d6ce96f2dc74e1c1e82931965dcadd